README
X-CART README
=============

Welcome!
This file contains basic information about X-Cart. More detailed information 
is available in X-Cart User manual.

I.    SYSTEM REQUIREMENTS
II.   INSTALLATION
III.  COMMENTS FOR DEVELOPER
IV.   X-CART STRUCTURE
V.    X-CART FILES
VI.   DEFAULT ACCOUNTS
VII.  SETTING UP AND CONFIGURING
VIII. BUG REPORTS


I. SYSTEM REQUIREMENTS
======================

For detailed information  on system requirements, refer to the User manual  for 
your X-Cart version. You can find it in the file area of the Support HelpDesk or
on our website (http://help.x-cart.com/index.php?title=X-Cart:User_manual_contents).
System requirements information is also available on this page:
http://help.x-cart.com/index.php?title=X-Cart:Server_Requirements_(latest_X-Cart_version)

Because of bugs in Opera v10.0, X-Cart will not work in this browser when an IP address is used in the store URL. Please use a domain name or upgrade/change your browser.

II. INSTALLATION
================

For detailed instructions on X-Cart installation, refer to the User manual for 
your X-Cart version.


III. COMMENTS FOR DEVELOPER
===========================

There are several ways to customize the look and feel of X-Cart:

1. You can modify the Smarty templates out of which the store's pages are built.
The templates can be edited using X-Cart's "Edit templates" or "Webmaster mode" 
features or any decent text editor.

2. You can modify the presentation of the store's pages (color, etc) by editing 
X-Cart's CSS files. These files can be edited using X-Cart's "Edit templates"
feature or any decent text editor.

3. You can upload your own images for use in the store's user interface to
the directory <xcart-dir>/skin1/images/ and get the store's templates to use 
them. The images can be uploaded using FTP or X-Cart's "Edit templates" feature. 
4. Most  of  X-Cart  texts  are  stored in "language variables".  You can  edit 
language  variables  under  the  "Languages" section  of  the  admin interface.

For information on Smarty, refer to the Smarty manual at 
http://www.smarty.net/manual/en/.

For detailed information on editing X-Cart's template and CSS files, as well as 
instructions for editing X-Cart's language variables, refer to the User manual 
for your X-Cart version.

To get a detailed consultation about modifying X-Cart, contact our support
service using your personal Support HelpDesk account at https://secure.qtmsoft.com.


IV. X-CART STRUCTURE
====================

1. PHP scripts:
   a) Customer related scripts are located in the ./ (root) directory.
   b) Administrator related scripts are located in the ./admin directory.
   c) Provider related scripts are located in the ./provider directory.
   d) Common scripts are located in the ./include directory.
   e) Modules related scripts are located in the ./modules directory.
   f) Customer/Administrator/Provider related scripts location is configured in
      the ./top.inc.php script.

2. Smarty class scripts are located in ./include/lib/smarty directory.

3. CSS styles are described in ./skin1/*.css file.

4. Smarty templates configuration can be found in ./skin1/skin1.conf file.

5. Templates directory:
   a) Common templates are located in the ./skin1 and ./skin1/main directories.
   b) Templates for the Customer area are located in the ./skin1/customer directory.
   c) Templates for the Admin area are located in the ./skin1/admin directory.
   d) Templates for the Provider area are located in the ./skin1/provider directory.
   e) Mail templates are located in the ./skin1/mail directory.
   f) Templates for X-Cart modules are located in the ./skin1/modules directory.

6. Compiled templates (for cache and Smarty internal use): ./var/templates_c

7. Dump of MySQL database is in ./sql/*.sql file.

8. Website images are located in ./skin1/images directory.

9. Product images, category images, manufacturer images etc are located in the 
./images directory.

10. Files related to integrated shipping modules are located in the ./shipping
   directory.


V. SETTING UP AND CONFIGURING
===============================

X-Cart system configuration is divided into several files.  

X-Cart's main configuration file - "config.php" -  is  located  under  main  
X-Cart  directory. This file contains variables for the following:

* SQL DB information
* Single Store operation mode flag
* Safe Administrator mode flag
* Customer titles
* Credit Card types accepted by your store
* etc

X-Cart's template configuration file is "skin1.conf". Configuration variables from
this file are used in templates and affect the look and feel of your store.


VI. BUG REPORTS
=================

You can report bugs and suggestions for improving X-Cart using your personal 
Support HelpDesk account at https://secure.qtmsoft.com.


                                 *  *  *  *  *

X-Cart
Copyright (c) 2001-2012 Qualiteam software Ltd <info@x-cart.com>
All rights reserved.



CHANGELOG
$Id: CHANGELOG,v 1.20.2.9.2.4 2012/04/23 06:59:31 aim Exp $

------------------ X-CART v4.5.0 RELEASED

*SHIPPING*

[*] 16 Apr 2012, aim - Improvement (0120653): Changed the 'Number of grams in the unit of weight defined by the weight symbol' from '453.6' to '453.59237'.

*PAYMENTS*

[!] 16 Apr 2012, aim - Bug (0120280): PayPal PRO Direct Payment fails to work with X-Cart via X-Payments. Fixed.
[!] 17 Apr 2012, aim - Bug (0120538): eWay merchant hosted payments error: XML not formatted correctly. Reason: Whitespace is not allowed. Fixed.
[*] 18 Apr 2012, aim - Improvement (0120787): Changed schema for the xcart_cc_pp3_data table: sessid char(32) NOT NULL DEFAULT ''.

*TAXES*

[*] 16 Apr 2012, aim - Improvement (0120552): Changes related to TaxCloud module.

*PRODUCTS*

[*] 11 Apr 2012, aim - Improvement (0113856): Added 'Compatibility with old product storage routine' setting.
[!] 20 Apr 2012, aim - Bug (0120314): Products Map fails to work with UTF-8. Fixed.

*IMPORT/EXPORT*

[*] 11 Apr 2012, aim - Improvement (0120543): Added data_type binary(3) field to xcart_import_cache.

*USERS*

[!] 16 Apr 2012, ferz - Bug (0120527): Address book: The state field remains in the profile after selecting a country without states. Fixed.

*MODULES/ADD-ONS*

[!] 13 Apr 2012, ferz - Bug (0120606): Froogle: Wholesale price used in some cases instead of original price. Fixed.
[!] 16 Apr 2012, ferz - Bug (0120611): XML Sitemap: Clean URLs for categories generated with missing trailing '/'. Fixed.
[!] 17 Apr 2012, aim - Bug (0120530): Antifraud problems: "declined_orders" and "completed_orders" calculated incorrectly. Fixed.
[!] 18 Apr 2012, aim - Bug (0120551): Expired cache not removed in the TaxCloud module. Fixed.
[!] 18 Apr 2012, aim - Bug (0120793): Cache not removed for the Amazon_Checkout module. Fixed.

*PERFORMANCE*

[*] 18 Apr 2012, aim - Improvement (0120785): Changed schema for the xcart_amazon_data table for Amazon_Checkout.

*MISCELLANEOUS*
[!] 02 Apr 2012, aim - Bug (0113021): 'I accept the "Terms & Conditions"' link fails to work when the shop closes on the partner registration page; 'X-Cart news' section fails to appear. Fixed.
[*] 16 Apr 2012, aim - Improvement (0120597): Updated COMPATIBLE_VERSION constant to 4.5.x.
[*] 16 Apr 2012, aim - Improvement (0120669): Moved help.qtmsoft.com to help.x-cart.com.
[!] 17 Apr 2012, aim - Bug (0120708): Database integrity issue. Fixed.
[*] 17 Apr 2012, aim - Improvement (0120328): Changed minimum MySQL version to 4.1.2.
[!] 18 Apr 2012, aim - Bug (0120789): Typo in the db_connection function. Fixed.
[!] 19 Apr 2012, aim - Bug (0106914): Admin top menu fails to work in iOS Safari browser (iPad/iPod). Fixed.

----------------- X-CART v4.5.0 beta 1 RELEASED

*CART*

[!] 29 Feb 2012, aim - Bug (0119066): Online payment methods fail to appear in the customer area. Fixed.

*SHIPPING*

[*] 07 Mar 2012, aim - Improvement (0102330): Added new 'Split the shipment into multiple packages if its weight/dimensions exceed the limitations' setting for the USPS module.
[*] 07 Mar 2012, aim - Improvement (0118585): Updated the USPS module.
[*] 07 Mar 2012, aim - Improvement (0118131): USPS server does not support the testing.shippingapis.com URL. The production.shippingapis.com URL is hardcoded for the USPS module.
[*] 07 Mar 2012, aim - Improvement (0117638): Added new methods for the USPS module.
[*] 07 Mar 2012, aim - Improvement (0116832): Added new Priority Mail Regional Box A/B/C containers for the USPS module.
[*] 09 Mar 2012, aim - Improvement (0068722): Removed depricated shipping methods for the USPS module.
[*] 09 Mar 2012, aim - Improvement (0092331): Corrected weight limits for First Class methods for the USPS module.
[*] 09 Mar 2012, aim - Improvement (0117303): Code refactoring for the USPS module. Removed global variable. Added functions: func_USPS_has_response_error, func_USPS_parse_methods, func_usps_save_pack_configuration_slg, func_USPS_prepare_rate_query, func_USPS_log_debug_information.
[*] 12 Mar 2012, aim - Improvement (0117303, 0119371): Adjusted func_usps_check_shippingid, related to the shipping label generator module of the USPS module.
[*] 12 Mar 2012, aim - Improvement (0108307): 'Value of content' is now related to order total for the USPS module.
[*] 12 Mar 2012, aim - Improvement (0117303): Fixed the 'Invalid First Class Mail Type' error related to the USPS module.
[*] 12 Mar 2012, aim - Improvement (0117303): Fixed the 'GIFT CARD FLAT RATE is an invalid container type for a REGULAR package and PRIORITY service' error in the USPS module.
[*] 13 Mar 2012, aim - Improvement (0117303): Fixed the 'REGIONALRATEBOXB is an invalid container type for a REGULAR package and PRIORITY service' error in the USPS module.
[*] 13 Mar 2012, aim - Improvement (0117303): Fixed the 'Service not available' error in the USPS module.
[*] 13 Mar 2012, aim - Improvement (0117303): Added new methods to the USPS module.
[*] 13 Mar 2012, aim - Improvement (0117303): Adjusted default settings for the USPS module.

*PAYMENTS*

[*] 02 Feb 2012, aim - Improvement (0118206): Changed PSiGate payment integration to the 'Interac Online' method (not involving credit cards).
[*] 02 Feb 2012, aim - Improvement (0117648): Changed background for 'eSec - ReDirect' and 'DIBS' to 'Y', as it will be removed.
[*] 06 Feb 2012, aim - Improvement (0117648): Added pop-up info window for deleted background payments methods for PCI-DSS compatibility.
[*] 15 Feb 2012, aim - Improvement (0117652): Removed the 'Add CC manual processing' feature from the payment methods page.
[*] 16 Feb 2012, aim - Improvement (0117652): Changed new payment method wizard.
[*] 17 Feb 2012, aim - Improvement (0117652): 'Phone Ordering' is now the default method.
[*] 17 Feb 2012, aim - Improvement (0117652): Adjusted the func_cart_add_offline_payment/func_get_offline_payment_methods tests.

cart.php
<?php
/* vim: set ts=4 sw=4 sts=4 et: */
/*****************************************************************************\
+-----------------------------------------------------------------------------+
| X-Cart Software license agreement                                           |
| Copyright (c) 2001-2012 Qualiteam software Ltd <info@x-cart.com>            |
| All rights reserved.                                                        |
+-----------------------------------------------------------------------------+
| PLEASE READ  THE FULL TEXT OF SOFTWARE LICENSE AGREEMENT IN THE "COPYRIGHT" |
| FILE PROVIDED WITH THIS DISTRIBUTION. THE AGREEMENT TEXT IS ALSO AVAILABLE  |
| AT THE FOLLOWING URL: http://www.x-cart.com/license.php                     |
|                                                                             |
| THIS AGREEMENT EXPRESSES THE TERMS AND CONDITIONS ON WHICH YOU MAY USE THIS |
| SOFTWARE PROGRAM AND ASSOCIATED DOCUMENTATION THAT QUALITEAM SOFTWARE LTD   |
| (hereinafter referred to as "THE AUTHOR") OF REPUBLIC OF CYPRUS IS          |
| FURNISHING OR MAKING AVAILABLE TO YOU WITH THIS AGREEMENT (COLLECTIVELY,    |
| THE "SOFTWARE"). PLEASE REVIEW THE FOLLOWING TERMS AND CONDITIONS OF THIS   |
| LICENSE AGREEMENT CAREFULLY BEFORE INSTALLING OR USING THE SOFTWARE. BY     |
| INSTALLING, COPYING OR OTHERWISE USING THE SOFTWARE, YOU AND YOUR COMPANY   |
| (COLLECTIVELY, "YOU") ARE ACCEPTING AND AGREEING TO THE TERMS OF THIS       |
| LICENSE AGREEMENT. IF YOU ARE NOT WILLING TO BE BOUND BY THIS AGREEMENT, DO |
| NOT INSTALL OR USE THE SOFTWARE. VARIOUS COPYRIGHTS AND OTHER INTELLECTUAL  |
| PROPERTY RIGHTS PROTECT THE SOFTWARE. THIS AGREEMENT IS A LICENSE AGREEMENT |
| THAT GIVES YOU LIMITED RIGHTS TO USE THE SOFTWARE AND NOT AN AGREEMENT FOR  |
| SALE OR FOR TRANSFER OF TITLE. THE AUTHOR RETAINS ALL RIGHTS NOT EXPRESSLY  |
| GRANTED BY THIS AGREEMENT.                                                  |
+-----------------------------------------------------------------------------+
\*****************************************************************************/

/**
 * This script implements shopping cart facility
 *
 * @category   X-Cart
 * @package    X-Cart
 * @subpackage Customer interface
 * @author     Ruslan R. Fazlyev <rrf@x-cart.com>
 * @copyright  Copyright (c) 2001-2012 Qualiteam software Ltd <info@x-cart.com>. All rights reserved
 * @license    http://www.x-cart.com/license.php X-Cart license agreement
 * @version    $Id: cart.php,v 1.211.2.20 2012/03/27 11:18:13 aim Exp $
 * @link       http://www.x-cart.com/
 * @see        ____file_see____
 */

//$_COOKIE["xid_793fc"]=null;

require './auth.php';



if (!empty($active_modules['Wishlist'])) {

    require $xcart_dir . '/modules/Wishlist/check_user.php';
    
}

x_load(
    'cart',
    'product'
);

x_session_register('cart');

/**
 * variables with integer type 
 */
foreach (
    array(
        'paymentid', 
        'transactionid', 
        'productid', 
        'shippingid', 
        'amount',
    ) as $int_var
) {
    if(isset($$int_var)) {
        $$int_var = intval($$int_var);
    }
}

if (
    'transaction_remove' === $mode
    && isset($cart['split_query'])
    && isset($cart['split_query']['transaction_query'][$paymentid][$transactionid])
) {

    $transaction = $cart['split_query']['transaction_query'][$paymentid][$transactionid];

    $payment_name = func_query_first_cell('SELECT processor FROM ' . $sql_tbl['ccprocessors'] . ' WHERE paymentid = \'' . $paymentid . '\'');
    $payment_name = preg_replace("/\.php/Ss", '', $payment_name);

    func_pm_load($payment_name);

    if (
        !empty($transaction['functions']['void'])
        && function_exists($transaction['functions']['void'])
    ) {

        $return = call_user_func_array($transaction['functions']['void'], $transaction['functions']['params_void']);

        // Void function MUST return TRUE value to remove transaction from the query
        if (true === $return) {

            unset($cart['split_query']['transaction_query'][$paymentid][$transactionid]);

            func_store_split_checkout_data($cart['split_query']);

        }

    }

    func_header_location('cart.php?mode=checkout&paymentid=' . $paymentid_return);
}

/**
 * Check if the cart is empty
 */
$func_is_cart_empty = func_is_cart_empty($cart);

require $xcart_dir . '/include/checkout_init.php';

/**
 * Normalize cart content
 */
if (
    !$func_is_cart_empty
    && !in_array($mode, array('wishlist', 'wl2cart'))
) {
    $cart_changed = func_cart_normalize($cart);
}

/**
 * Clear entire cart
 */
if ($mode == 'clear_cart') {

    $changes = array();

    if (!empty($cart['products'])) {

        foreach($cart['products'] as $p) {

            $changes[$p['cartid']] = array(
                'productid' => $p['productid'],
                'quantity'  => $p['amount'],
                'change'    => $p['amount'] * (-1),
            );

        }

    }

    if (!empty($cart['giftcerts'])) {

        foreach($cart['giftcerts'] as $gid => $g) {

            $changes[$gid] = array(
                'gcindex'  => $gid,
                'quantity' => 1,
                'change'   => -1,
            );

        }

    }

    if (
        !empty($active_modules['SnS_connector'])
        && !empty($cart['products'])
    ) {

        foreach ($cart['products'] as $p) {

            $is_sns_action['DeleteFromCart'][] = $p['productid'];

        }

    }

    $cart = '';

    func_register_ajax_message(
        'cartChanged',
        array(
            'changes' => $changes,
            'isEmpty' => true,
            'status'  => 1
        )
    );

    $cart_checksum = '';
    func_header_location('cart.php');
}

/**
 * Unset Gift Certificate
 */
if (
    !empty($active_modules['Gift_Certificates'])
    && $mode == 'unset_gc'
    && $gcid
) {
    func_giftcert_unset($gcid);

    func_register_ajax_message(
        'opcUpdateCall',
        array(
            'action'   => 'updateGC',
            'status'   => 1,
            'gc_total' => $cart['giftcert_discount']
        )
    );

    if (
        func_is_ajax_request()
        && func_cart_is_payment_methods_list_changed(@$payment_methods)
    ) {
        func_register_ajax_message(
            'opcUpdateCall',
            array(
                'action' => 'paymentMethodListChanged'
            )
        );
    }

    func_header_location('cart.php?mode=checkout' . ($paymentid ? '&paymentid=' . $paymentid : ''));
}

/**
 * Update Gift wrapping
 */
include $xcart_dir . '/modules/Gift_Registry/giftreg_customer_cart.php';

$smarty->assign(
    'register_script_name',
    (
        ($config['Security']['use_https_login'] == 'Y')
            ? $xcart_catalogs_secure['customer'] . '/'
            : ''
    )
    . 'cart.php'
);

$return_url = '';

/**
 * Add product to cart
 */
if (
    $mode == 'add'
    && !empty($productid)
) {
    $add_product = array(
        'productid'       => abs($productid),
        'amount'          => abs($amount),
        'product_options' => isset($product_options) ? $product_options : array(),
        'price'           => isset($price) ? func_convert_number($price) : null,
    );

    if (!empty($active_modules['Special_Offers'])) {

        include $xcart_dir . '/modules/Special_Offers/add_to_cart.php';

    }

    $result = func_add_to_cart($cart, $add_product);

    if (
        !isset($_GET['redirect_to_referer'])
        && !empty($result['redirect_to'])
    ) {
        func_header_location($result['redirect_to']);
    }

    // Recalculate cart totals after new item added
    list($cart, $products) = func_generate_products_n_recalculate_cart();

    if (!empty($active_modules['Special_Offers'])) {

        include $xcart_dir . '/modules/Special_Offers/add_free_products.php';

    }

    if (isset($_GET['redirect_to_referer'])) {

        func_header_location($HTTP_REFERER . '&redirect_from_cart=Y');

    }

    $func_is_cart_empty = func_is_cart_empty($cart);

    func_register_ajax_message(
        'cartChanged',
        array(
            'changes'     => array(
                $result['productindex'] => array(
                    'productid' => $add_product['productid'],
                    'quantity'  => $result['quantity'],
                    'changed'   => $result['changed'],
                )
            ),
            'isEmpty' => empty($cart['products']) && empty($cart['giftcerts']),
            'status'  => $result['status'],
        )
    );

    // Redirect
    if (
        $config['General']['redirect_to_cart'] == 'Y'
        || isset($_GET['redirect_to_cart'])
    ) {

        if (!empty($active_modules['SnS_connector']))
            $is_sns_action['AddToCart'][] = $productid;

        $return_url = 'cart.php';

    } else {

        list($cart, $products) = func_generate_products_n_recalculate_cart(); 

        if (!empty($active_modules['SnS_connector']))
            func_generate_sns_action('AddToCart', $productid);

        func_save_customer_cart();

        if (func_is_internal_url($HTTP_REFERER)) {

            $tmp = @parse_url($HTTP_REFERER);
            $return_url = $HTTP_REFERER;

            if (
                $config['General']['return_to_dynamic_part'] == 'Y'
                && !empty($is_submit_from_html_page)
                && is_array($tmp)
                && (
                    strpos($tmp['path'], '.html') !== false
                    || substr($tmp['path'], -1) == '/'
                )
            ) {

                if (substr($tmp['path'], -1) == '/') {

                    $return_url = 'home.php';

                } elseif (strpos($HTTP_REFERER, '-c-') !== false) {

                    $return_url = func_get_resource_url('category', $cat, 'page=' . $page);

                } else {

                    $return_url = func_get_resource_url('product', $add_product['productid']);

                }

            }

        } elseif(!empty($cat)) {

            $return_url = func_get_resource_url('category', $cat, 'page=' . $page);

        }

    }

}

/**
 * Delete product from the cart
 */
if (
    $mode == 'delete'
    && !empty($productindex)
) {
    $productid = 0;
    $quantity = 0;

    if (
        !empty($cart['products'])
        && is_array($cart['products'])
    ) {
        list($productid, $quantity) = func_delete_from_cart($cart, $productindex);
    }

    if ($productid > 0) {

        if (!empty($active_modules['SnS_connector']))
            $is_sns_action['DeleteFromCart'][] = $productid;

        // Recalculate cart totals after updating
        list($cart, $products) = func_generate_products_n_recalculate_cart();
    }

    func_register_ajax_message(
        'cartChanged',
        array(
            'changes' => array(
                $productindex   => array(
                    'productid' => $productid,
                    'quantity'  => $quantity,
                    'changed'   => $quantity * -1
                )
            ),
            'isEmpty' => empty($cart['products']) && empty($cart['giftcerts']),
            'status'  => $productid > 0 ? 1 : 2
        )
    );

    $func_is_cart_empty = func_is_cart_empty($cart);

    $return_url = 'cart.php';
}

if (empty($action)) {

    $action = '';

}

$changes = false;

/**
 * Update shopping cart
 */
if (
    $action == 'update'
    && !$func_is_cart_empty
) {
    // Update Gift registry links, if any
    if (!empty($active_modules['Gift_Registry'])) {

        include $xcart_dir . '/modules/Gift_Registry/giftreg_customer_cart.php';

    }

    // Update quantity of products in cart
    if (!empty($productindexes)) {

        list($min_amount_warns, $changes) = func_update_quantity_in_cart($cart, $productindexes);

        $top_message = func_generate_min_amount_warning($min_amount_warns, $productindexes, $cart['products']);

        if (!empty($active_modules['SnS_connector']))
            $is_sns_action['CartChanged'][] = false;

        $intershipper_recalc = 'Y';
    }

    // Update shipping method
    if (
        $config['Shipping']['realtime_shipping'] == 'Y'
        && !empty($active_modules['UPS_OnLine_Tools'])
        && $config['Shipping']['use_intershipper'] != 'Y'
        && isset($selected_carrier)
    ) {
        $current_carrier = $selected_carrier;

        if (empty($shippingid)) {
            // Reset active COD payment method if COD shipping is changed to nonCOD shipping
            $is_cod_disabled_payment = 
                func_query_first_cell("SELECT is_cod FROM $sql_tbl[shipping] WHERE shippingid='$cart[shippingid]'") == 'Y' &&
                func_query_first_cell("SELECT is_cod FROM $sql_tbl[payment_methods] WHERE paymentid='$cart[paymentid]'") == 'Y';
            
            if ($is_cod_disabled_payment) {
                $cart = func_cart_set_paymentid($cart, 0);
                $paymentid = 0;
            }    

        }    
    }

    if (
        !empty($shippingid)
        && $cart['shippingid'] != $shippingid
    ) {

        $cart = func_cart_set_shippingid($cart, $shippingid);

        func_register_ajax_message(
            'opcUpdateCall',
            array(
                'action' => 'shippingChanged',
                'value'  => $shippingid
            )
        );
    }

    if (!empty($cart['is_payment_changed'])) {

        $cart['is_payment_changed'] = false;

        func_register_ajax_message(
            'opcUpdateCall',
            array(
                'action' => 'paymentChanged',
                'value'  => $paymentid
            )
        );
    }

    if (!empty($mode)) {

        $url_args[] = 'mode=' . $mode;

    }

    if (
        !empty($paymentid)
        && $checkout_module != 'One_Page_Checkout' // Paymentid is stored in cart['paymentid'] for One_Page_Checkout
    ) {

        $url_args[] = 'paymentid=' . $paymentid;

    }

    $return_url = 'cart.php' . (!empty($url_args) ? '?' . implode('&', $url_args) : '');

    func_register_ajax_message(
        'cartChanged',
        array(
            'changes' => is_array($changes) ? $changes : array(),
            'isEmpty' => empty($cart['products']) && empty($cart['giftcerts']),
            'status'  => 1,
        )
    );

    $func_is_cart_empty = func_is_cart_empty($cart);
}

// Prepare the products data

$products = func_products_in_cart($cart, @$userinfo['membershipid']);

/**
 * Apply / unset discount coupon
 */

if (!empty($active_modules['Discount_Coupons'])) {

    include $xcart_dir . '/modules/Discount_Coupons/discount_coupons.php';

}

/**
 * Calculate totals
 */

include $xcart_dir . '/include/cart_calculate_totals.php';

if (!empty($active_modules['SnS_connector'])) {
    func_sns_exec_actions($is_sns_action);
}

if (
    isset($totals_checksum)
    && $totals_checksum_init !== $totals_checksum
) {

    func_register_ajax_message(
        'opcUpdateCall',
        array(
            'action' => 'updateTotals'
        )
    );

    if (
        func_is_ajax_request()
        && func_cart_is_payment_methods_list_changed(@$payment_methods)
    ) {
        func_register_ajax_message(
            'opcUpdateCall',
            array(
                'action' => 'paymentMethodListChanged'
            )
        );
    }

}

func_save_customer_cart();

if ($return_url) {

    func_header_location($return_url);

}

/**
 * Wishlist facility
 */
if (
    !empty($active_modules['Wishlist'])
    && $mode != 'checkout'
) {
    if (
        $mode == 'move_product'
        && !empty($active_modules['Gift_Registry'])
    ) {
        include $xcart_dir . '/modules/Gift_Registry/giftreg_wishlist.php';
    }

    include $xcart_dir . '/modules/Wishlist/wishlist.php';
}

if (
    $mode != 'wishlist'
    || empty($active_modules['Wishlist'])
) {
    $location[] = ($mode == 'checkout')
        ? array(func_get_langvar_by_name('lbl_checkout'), '')
        : array(func_get_langvar_by_name('lbl_your_shopping_cart'), '');
}

// Include common scripts
include $xcart_dir . '/include/common.php';

$giftcerts = !empty($cart['giftcerts'])
    ? $cart['giftcerts']
    : array();

// Update minicart
include $xcart_dir . '/minicart.php';

if (
    isset($userinfo)
    && !empty($userinfo)
) {
    $smarty->assign('userinfo', $userinfo);
}

if (!$func_is_cart_empty) {
    $smarty->assign('products',        $products);
    $smarty->assign('giftcerts',       $giftcerts);
    $smarty->assign('list_length',     count($products) + count($giftcerts));
    $smarty->assign('products_length', count($products));
}

func_save_customer_cart();

if (
    !$func_is_cart_empty
    && $mode == 'checkout'
) {
    require $xcart_dir . '/include/checkout.php';
}

x_session_save();

if (
    !isset($main)
    || empty($main)
) {
    $main = 'cart';
}

$smarty->assign('main', $main);

if (!$func_is_cart_empty) {

    if ($main == 'cart' || $main == 'checkout') {

        $page_container_class .= ' ' . 'checkout-container';
        $smarty->assign('page_container_class', $page_container_class);

        if ($main != 'checkout') {
            $smarty->assign('shipping_estimate_fields', $shipping_estimate_fields);
        }
    }

} else {

    $smarty->assign('cart_empty', true);

}

// Assign the current location line
$smarty->assign('location', $location);

func_display('customer/home.tpl', $smarty);

?>
